### ISSUE TITLE
Compile on Ubuntu 14.04

### ISSUE THREAD

**Issue by probonopd** [solvespace/solvespace#344]
### System information

master

Operating system: Ubuntu 14.04

### Expected behavior

Can compile the GUI

### Actual behavior

https://github.com/solvespace/solvespace/blob/2b4c484dcb1031f9f196805def71887b92ab097f/.travis/build-debian.sh#L7

### Additional information

We need to find a ppa or other software source that provides a suitable version of Gtk+. Alternatively, we need to lower the requirements for Solvespace so that it can be compiled using the Gtk+ that comes with Ubuntu 14.04, the currently oldest still-supported LTS version.
I consider it good practice to target the oldest still-supported LTS versions of distributions when developing applications.

**Comment by whitequark** [comment]
> Alternatively, we need to lower the requirements for Solvespace so that it can be compiled using the Gtk+ that comes with Ubuntu 14.04, the currently oldest still-supported LTS version.

This will not happen. Any GTKmm version under 3.16 (including GTK+ 2; there is a reason I don't support it anymore) has fundamentally unusable OpenGL support, and may not be used with SolveSpace. Actually, 3.16 is already supported using a number of workarounds; the original requirement was 3.22.

The GTK 3 release with stable ABI is 3.22, so I cannot and will not guarantee that SolveSpace works with any GTK 3 release that isn't 3.22 or later. If you think this is unacceptable, well, you can complain to GTK developers about their terrible release schedule.

> We need to find a ppa or other software source that provides a suitable version of Gtk+.

As discussed earlier, I have decided that packaging for extremely old Linux distro versions is out of scope of the SolveSpace project. Moreover, considering earlier discussions on shared libraries, theming, and user experience, and the fact that theming became stable only in GTK+ 3.22, I cannot endorse any AppImage except one that:
  * uses GTK+ 3.22 or later, and
  * uses theming plugins from the target system.

Such an AppImage could, in principle, carry its own copy of GTK+, but considering that using theming plugins from the target system requires that it would have a copy of GTK+ 3.22 *anyway*, it probably should just dynamically load GTK+ from the target system.

In other words, I don't think AppImages built against GTK+ earlier than 3.22 should be made. Of course, I cannot prevent you from doing so, but I do not think it is a good idea and I do not support this use case.

In terms of Ubuntu releases, this excludes both Trusty (which shipped with 3.10) and Xenial (which shipped with 3.18), leaving only Bionic (which finally shipped with 3.22.2).

**Comment by probonopd** [comment]
> In terms of Ubuntu releases, this excludes both Trusty (which shipped with 3.10) and Xenial (which shipped with 3.18), leaving only Bionic (which finally shipped with 3.22.2).

Well, _you_ are the developer of the application, but this is a bit like requiring Windows 10 or macOS 10.13 and leaving everyone else out...

**Comment by whitequark** [comment]
> Well, you are the developer of the application, but this is a bit like requiring Windows 10 or macOS 10.13 and leaving everyone else out...

It's funny that you mention requiring these particular versions of these operating systems, because Windows 7 is long past its end of support (and with security updates through 2020, it has nearly reached the end of its useful life as well), and requiring the latest version of macOS is commonplace and accepted in that ecosystem. The only reason I *don't* require Windows 10 and macOS 10.13 is because those operating systems actually take compatibility seriously, unlike GTK+, and it doesn't cost me anything to keep supporting Windows Vista and macOS 10.9.

To reiterate: if you think the current situation is bad, you can complain to the GTK+ developers and ask them to provide the same ABI stability guarantees as every other OS and most GUI toolkits (such as Qt) available.

**Comment by whitequark** [comment]
So, as I've explained, Trusty will never be officially supported. Xenial *is* supported--you can compile SolveSpace on Xenial without PPAs--but whether an AppImage built against Xenial would work against Bionic is an open question. If such an AppImage can be dynamically linked against, and not redistribute GTK, then it is an AppImage I can integrate into our nightly build process, likely via Docker. If not, well, the AppImage will be only for Bionic and newer distributions.

Closing this in favor of #347.

**Comment by probonopd** [comment]
> whether an AppImage built against Xenial would work against Bionic is an open question

I assume that a binary built on an older operating system should _always_ run on a _newer_ version of the operating system, or else it's a bug. This has nothing to do with AppImage specifically, it's just how backward compatibility works (which we should expect from OSes).

**Comment by whitequark** [comment]
> This has nothing to do with AppImage specifically, it's just how backward compatibility works (which we should expect from OSes).

Yes, we should, but GTK doesn't guarantee that a binary built against GTK 3 <3.22 will work against GTK 3 >=3.22. As I've been saying many times, the GTK developers either don't understand how backwards compatibility works literally everywhere else, or just don't care.

**Comment by probonopd** [comment]
> GTK doesn't guarantee that a binary built against GTK 3 <3.22 will work against GTK 3 >=3.22

Which equals, at least for me, to "GTK 3 <3.22 is broken and should never have been named 3x".

Great example that it is _not_ "OK" that Gtk+ 3.00 is not "Gtk+ 3". Unfortunately they seem to be repeating their old mistakes time and again: 

* https://blogs.gnome.org/desrt/2016/06/13/gtk-4-0-is-not-gtk-4/
* https://blogs.gnome.org/desrt/2016/06/14/gtk-5-0-is-not-gtk-5/

**Comment by whitequark** [comment]
> Which equals, at least for me, to "GTK 3 <3.22 is broken and should never have been named 3.x".

Sure. But combine this with the fact that GTK 2 is *also* broken, since it has no HiDPI support, which makes it unusable on such devices, and you realize that I have no choice but to target one of the broken versions. Well, that, or migrate to Qt, which seems like a better option with every day wasted trying to do something useful with GTK.

**Comment by aruiz** [comment]
> Yes, we should, but GTK doesn't guarantee that a binary built against GTK 3 <3.22 will work against GTK 3 >=3.22. As I've been saying many times, the GTK developers either don't understand how backwards compatibility works literally everywhere else, or just don't care.

Wrong. There will be no further stable 3.x releases beyond 3.22 (3.9x will be unstable development releases not meant for production and backwards compatibility will be guaranteed at 4.0 time).

**Comment by whitequark** [comment]
@aruiz Do I understand it correctly that GTK dropped the versioning scheme described in [this post](https://blogs.gnome.org/desrt/2016/06/13/gtk-4-0-is-not-gtk-4/)?

**Comment by ebassi** [comment]
> Yes, we should, but GTK doesn't guarantee that a binary built against GTK 3 <3.22 will work against GTK 3 >=3.22.

Of course this is nonsense: a binary built with GTK+ < 3.22 is guaranteed to run against GTK+ ≥ 3.22. The only thing that could possibly have changed are the CSS selectors that you may use for custom styling; different minor versions up to GTK+ 3.20 (when the CSS selectors were finally exposed as public API) can be namespaced, and you can ship different copies if you want to support different versions of GTK+.

> As I've been saying many times, the GTK developers either don't understand how backwards compatibility works literally everywhere else, or just don't care.

Or maybe you've just not bothered to read the change summary and notes for every release. Or bothered to talk to the GTK+ developers. Or maybe you just enjoy insulting people and their work.

**Comment by ebassi** [comment]
> @aruiz Do I understand it correctly that GTK dropped the versioning scheme described in this post?

Yes; that blog was superceded by the official stance of the GTK+ team as published [on the GTK+ development blog](https://blog.gtk.org/2016/09/01/versioning-and-long-term-stability-promise-in-gtk/), with the amendment of [a GTK+ 3.24](https://blog.gtk.org/2018/06/23/a-gtk-3-update/) release added only for ease of porting to GTK+ 4.0, and which does not break any of the stability promises.

**Comment by whitequark** [comment]
@ebassi 

> Of course this is nonsense: a binary built with GTK+ < 3.22 is guaranteed to run against GTK+ ≥ 3.22.

The blog post I linked before is specifically stating that minor releases will *not* provide ABI compatibility:

> Each 6 months, the new release (Gtk 4.2, Gtk 4.4, Gtk 4.6) will break API and ABI vs. the release that came before it

... and implies that this applies to the 3.x release series:

> The first “API stable” version under this new scheme is likely to be something like Gtk 3.26.

So I think you can see why I would get the impression that a binary built against GTK+ <3.22 isn't guaranteed to run against GTK+ >=3.22. It also doesn't help that a blog post that was the primary source of information on the new versioning scheme back when it was introduced is not updated with the information that what it describes is no longer valid.

But I'm glad that in the end I'm wrong, and GTK+ does, in fact, provide proper backwards compatibility guarantees. This means that we can provide AppImages that work on Xenial and later.

**Comment by probonopd** [comment]
So @ebassi I'll happily stand corrected. This sounds much better than what I read in those "personal blogs".

**Comment by Symbian9** [comment]
> But I'm glad that in the end I'm wrong, and GTK+ does, in fact, provide proper backwards compatibility guarantees. This means that we can provide AppImages that work on Xenial and later.

Sounds good, @probonopd ?

- https://github.com/solvespace/solvespace/issues/347

P.S.: Maybe, using backports of all needed components (there are many PPA's with backports of common libs for Trusty), it would be possible also create AppImage that will work on Trusty 14.04 too.

**Comment by Symbian9** [comment]
> Sadly, personal blogs are a thing; that's why the new versioning scheme was published on the development blog for GTK+, as a verified source of information.

@ebassi, you mean this?

- https://blog.gtk.org/2016/09/01/versioning-and-long-term-stability-promise-in-gtk/

**Comment by probonopd** [comment]
Yes, sounds good indeed. I was mislead by some private blogs of what seemed like a Gtk+ developer. As I wrote above, i think this is a sane policy.

**Comment by whitequark** [comment]
The 3.16 (or 3.18, I don't remember exactly) requirement is due to GtkGLArea not being exposed via gtkmm in earlier gtkmm versions. It is present in GTK+, just not in gtkmm.
